### PR TITLE
[FIX] core: update on new records should only recompute new records

### PR DESCRIPTION
**Pull request by rco-odoo**
The following situation can occur: a computed field depends on one2many
lines, and during an onchange, a line is modified.  Both the main record
and the line are new records.  If the many2one field is not in cache on
the line, its value is fetched from the line's origin, and the field
will be triggered for recomputation on the main record's origin instead
of the record itself.

The fix enforces the following property: if new records are modified,
only new records can be triggered for recomputation.

### CODE REVIEW

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo retry

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by xmo-odoo** [comment]
@robodoo retry

File: odoo/fields.py [hunk 1]
```diff
@@ -986,7 +986,12 @@ def __get__(self, record, owner):
         if self.compute and (record.id in env.all.tocompute.get(self, ())) \
                 and not env.is_protected(self, record):
             # self must be computed on record
-            recs = record if self.recursive else env.records_to_compute(self)
+            if self.recursive:
+                recs = record
+            else:
+                # do not mix real records with new records
```
Comment (xmo-odoo) [inline odoo/fields.py:992]: comment should be above the filtered, and maybe should more clearly note that we only want to compute records with the same "newness" as the reference, as at first glance it kind-of gives the impression that we're *partitioning* the records

File: odoo/sql_db.py [hunk 2]
```diff
@@ -60,12 +60,14 @@ def adapt_string(adapted):
 from inspect import currentframe
 
 
-def flush_env(cr):
+def flush_env(cr, clear=True):
```
Comment (xmo-odoo) [inline odoo/sql_db.py:63]: Might be better to not have a default value here. And it should probably be a keyword-only parameter either way.

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

File: odoo/sql_db.py (hunk 2 above)
Comment (rco-odoo) [reply odoo/sql_db.py:63]: This is a breaking API change, and we are on a stable version here.  So the default is necessary.

**Comment by rco-odoo** [comment]
@robodoo r+

**Pull request closed without merging**